slug.go

```diff
@@ -382,119 +392,92 @@ func (p *Packer) Unpack(r io.Reader, dst string) error {
 		if err := os.MkdirAll(dir, 0755); err != nil {
 			return fmt.Errorf("failed to create directory %q: %w", dir, err)
 		}
 
-		// Handle symlinks.
-		if header.Typeflag == tar.TypeSymlink {
+		// Handle symlinks, directories, non-regular files
+		if info.IsSymlink() {
 			if ok, err := p.validSymlink(dst, header.Name, header.Linkname); ok {
 				// Create the symlink.
-				if err = os.Symlink(header.Linkname, path); err != nil {
+				if err = os.Symlink(header.Linkname, info.Path); err != nil {
 					return fmt.Errorf("failed creating symlink (%q -> %q): %w",
 						header.Name, header.Linkname, err)
 				}
 			} else {
 				return err
 			}
 
+			if err := info.RestoreInfo(); err != nil {
+				return err
+			}
+
+			continue
+		}
+
+		if info.IsDirectory() {
+			// Restore directory info after all files are extracted because
+			// the extraction process changes directory's timestamps.
+			directoriesExtracted = append(directoriesExtracted, info)
 			continue
 		}
 
-		// Only unpack regular files from this point on.
-		if header.Typeflag == tar.TypeDir || header.Typeflag == tar.TypeXGlobalHeader || header.Typeflag == tar.TypeXHeader {
+		// The remaining logic only applies to regular files
+		if !info.IsRegular() {
 			continue
-		} else if header.Typeflag != tar.TypeReg && header.Typeflag != tar.TypeRegA {
-			return fmt.Errorf("failed creating %q: unsupported type %c", path,
-				header.Typeflag)
 		}
 
 		// Open a handle to the destination.
-		fh, err := os.Create(path)
+		fh, err := os.Create(info.Path)
 		if err != nil {
 			// This mimics tar's behavior wrt the tar file containing duplicate files
 			// and it allowing later ones to clobber earlier ones even if the file
-			// has perms that don't allow overwriting.
+			// has perms that don't allow overwriting. The file permissions will be restored
+			// once the file contents are copied.
 			if os.IsPermission(err) {
-				os.Chmod(path, 0600)
-				fh, err = os.Create(path)
+				os.Chmod(info.Path, 0600)
+				fh, err = os.Create(info.Path)
 			}
 
 			if err != nil {
-				return fmt.Errorf("failed creating file %q: %w", path, err)
+				return fmt.Errorf("failed creating file %q: %w", info.Path, err)
 			}
 		}
 
-		// Copy the contents.
+		// Copy the contents of the file.
 		_, err = io.Copy(fh, untar)
 		fh.Close()
 		if err != nil {
-			return fmt.Errorf("failed to copy slug file %q: %w", path, err)
+			return fmt.Errorf("failed to copy slug file %q: %w", info.Path, err)
 		}
 
-		// Restore the file mode. We have to do this after writing the file,
-		// since it is possible we have a read-only mode.
-		mode := header.FileInfo().Mode()
-		if err := os.Chmod(path, mode); err != nil {
-			return fmt.Errorf("failed setting permissions on %q: %w", path, err)
+		if err := info.RestoreInfo(); err != nil {
+			return err
+		}
+	}
+
+	// Now that extraction is complete, restore mode and timestamps previously saved
+	// about directories.
+	for _, dir := range directoriesExtracted {
```

Comment:
Are we guaranteed to have the directories occur in the right order here, or do we need to sort this list?

eg: `/foo/bar/` always before `/foo/`?

I know it would be possible to craft a file that would have a problem with this processing, and seeing the `os.MkdirAll` suggests that there aren't ordering guarantees.

Reply:
Good call out. I think this is only working based on how tree walk works during Pack.

Reply:
✅